MailCore/Cache/Actions/ActionsManager.swift

```diff
@@ -0,0 +1,245 @@
+/*
+ Infomaniak Mail - iOS App
+ Copyright (C) 2022 Infomaniak Network SA
+
+ This program is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ This program is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import Foundation
+import InfomaniakCoreUI
+import InfomaniakDI
+import MailResources
+import SwiftUI
+
+extension [Message]: Identifiable {
+    public var id: Int {
+        // Calculate a unique identifier by XORing hash values of messages
+        return reduce(1) { $0.hashValue ^ $1.hashValue }
+    }
+}
+
+extension RandomAccessCollection where Element == Message {
+    public func lastMessageToExecuteAction(currentMailboxEmail: String) -> Message? {
+        if let message = last(where: { $0.isDraft == false && $0.fromMe(currentMailboxEmail: currentMailboxEmail) == false }) {
+            return message
+        } else if let message = last(where: { $0.isDraft == false }) {
+            return message
+        }
+        return last
+    }
+
+    func lastMessagesAndDuplicatesToExecuteAction(currentMailboxEmail: String) -> [Message] {
+        let lastMessages = uniqueThreads.compactMap { $0.lastMessageToExecuteAction(currentMailboxEmail: currentMailboxEmail) }
+        return lastMessages + lastMessages.flatMap(\.duplicates)
+    }
+
+    func addingDuplicates() -> [Message] {
+        return self + flatMap(\.duplicates)
+    }
+
+    var uniqueThreads: [Thread] {
+        return Set(compactMap(\.originalThread)).toArray()
+    }
+
+    var isSingleMessage: Bool {
+        guard let firstMessageThreadMessagesCount = first?.originalThread?.messages.count else {
+            return false
+        }
+
+        return count == 1 && firstMessageThreadMessagesCount > 1
+    }
+}
+
+public class ActionsManager: ObservableObject {
+    @LazyInjectService private var snackbarPresenter: SnackBarPresentable
+
+    private let mailboxManager: MailboxManager
+    private let navigationState: NavigationState?
+
+    public init(mailboxManager: MailboxManager, navigationState: NavigationState?) {
+        self.mailboxManager = mailboxManager
+        self.navigationState = navigationState
+    }
+
+    public func performAction(target messages: [Message], action: Action, origin: ActionOrigin) async throws {
+        let messagesWithDuplicates = messages.addingDuplicates()
+
+        switch action {
```

Comment:
If we hit the max cyclomatic complexity, we probably should have a function for every action (some already have), or something like that.

Reply:
Yup totally agree

Reply:
Also refactored a bit in latest commits